### ISSUE TITLE
DynamicImage->ImageBuffer .. supportable types

### ISSUE THREAD

**Issue by dobkeratops** [PistonDevelopers/image#827]
http://www.piston.rs/image/image/struct.ImageBuffer.html

http://www.piston.rs/image/image/enum.DynamicImage.html

ok writing a texture loader from this.. at first , 24bit was fine. then generalising it to handle 24bit/32bit seemed slightly awkward (i might be a bit fussy about eliminating some cut-paste that happens because you need to match to get the buffer)

The effort will be worthwhile however if this wrapping(which seems to allow a type-parameter for the image buffer type) allows safe handling of a multitude of formats:

* 24,32bit - done
* 16bit:  1555,  565, 4444 - these do have some use-cases
* DXT compression .. 

I was thinking an easier way to make image buffers safe would be to use a 3D array as an intermediate: and moving images to & from **AI libraries** more frequently involves multi-dimensional array operations (especially consider 4D arrays for the feature detectors, and synergy between that and 'array texture' code, i.e. "16 slices of 4-channel RGBA 512x512 images" = u8[512][512][4][16]

.. but if this enum/match approach can handle different bit-level encodings, that would be great

**Comment by fintelia** [comment]
Would using the [`ImageDecoder` trait](https://docs.rs/image/0.20.0/image/trait.ImageDecoder.html) directly work better for your use case?

**Comment by dobkeratops** [comment]
@fintelia thanks for the info.. it's hard to tell, it seems to solve a different problem i.e. providing more interface to step through an image, whereas I just want to pass it to GL.
it also seems to decode then tell you a format via a variant meaning the usage code will look similar; I suspect in those kind of cases i'd want my code to ask for the image in a specific element format (i.e. "i'm going to do some image processing in HDR range fp32 ; give me the image as a 3D array of fp32 values" - because you've chosen to run on a machine that's biased to that format or you're doing somthing that needs that range for intermediates ).  (I was surprising impressed with how useable numpy was just throwing arrays around for these kind of use cases, and ditto *later* if i do move into AI-related image processing stuff i think thats what code would look like. I'm talking about 2 different things here.. just loading into an engine, and more advanced image processing)

**Comment by dobkeratops** [comment]
to clarify the 'problem' here, not so bad but it will get worse scaling for more texture formats:-

```
fn create_texture_from(dimg:image::DynamicImage)->TextureHandle{

	if let Some(((us,vs),bfr,fmt))=
		(match dimg {
			image::DynamicImage::ImageRgb8(img)=> Some((img.dimensions(),img.into_vec(),GL_RGB)),
			image::DynamicImage::ImageRgba8(img)=> Some((img.dimensions(),img.into_vec(),GL_RGBA)),
     ^^^^^^^^^^^^^^^^ // It's this repetition, cut-pasting code extracting the same information per variant
			_=> None
		})
	{unsafe {
```

**Comment by dobkeratops** [comment]
another way to do it with less repetition and 'ceremony' is to get a buffer with a set of numeric values:  num bits per channel, num channels,  width, height.  total buffer size, and an enum for channel format = (bits*channes*width*height)/8.  'bits per channel' will be 8.  or you could even say "bits per red, green,blue,alpha; width; height;  total size=((bits r+g+b)*w*h)/8", and that will handle all the packed formats. you still get to leverage Rusts' match statement in passing to GL but cleaner...

```
       let format = match (bits_r,bits_g,bits_b,bits_a) {
              (8,8,8,0) => (GL_RGB),
              (8,8,8,8) => (GL_RGBA),
            // etc
       }
(EDIT you can certainly add 'luma' to that list, the RGB's wil just match with a zero, and vica versa)

**Comment by fintelia** [comment]
I'm not sure I follow. Couldn't you do something like:

```rust
fn create_texture_from<D: ImageDecoder>(decoder: D)->Result<TextureHandle>{
    let (width, height) = decoder.dimensions()?;
    let format = match decoder.colortype()? {
        ColorType::RGB(8) => GL_RGB,
        ColorType::RGBA(8) => GL_RGBA,
        ColorType::Gray(8) => GL_LUMINANCE,
        ColorType::Palette(_) => unreachable!(),
        _ => unimplemented!(),
    };
    let contents = match decoder.read_image()? {
        DecodingResult::U8(v) => v,
        DecodingResult::U16(_) => unimplemented!(), // A bit sad to have this special case
    };

    // Actual create texture
}
```

**Comment by dobkeratops** [comment]
can't it just have something that returns a buffer and a numerical description of the bits as I indicated, then there's less matching ceremony to go through . you've got 2 matches there. 

I think the issue here is this library is designed for a broader range of use cases (i.e. actually writing conversions from one format to another) - fair enough. I'm just suggesting adding some peices that will streamline the most common tasks people are likely to encounter (grab an image in a desired format, or hand whats there to openGL)

**Comment by dobkeratops** [comment]
the nice thing about reporting it as numbers of bits is that some of the alternative code paths in other situations will be unified by just using those bit values to extract data with shifts and masks directly: less matching/wrapping, more direct computing, smaller code size, less documents to dig through. I think this is a win all round.

**Comment by dobkeratops** [comment]
anyway i'll give it a rest. i've got just 1 "offending" line, its ok. it does what I need right now.

I suppose i'm trying to think ahead , is there anything I could add to the library based on things i'm about to do (and things i could change that would make it better for future users)

One more thing to talk about is normal map formats 

and ok the "bits" idea does break down with compressed textures (so it would still need a match... "Uncompressesd() vs Compressed()"  with  the bits idea being in the 'uncompressed' arm.

**Comment by fintelia** [comment]
Yeah, the issue with just numbers of bits is that there are a bunch of other formats that can't be described that way, including compresesd images, images that have palettes, BGR images and so forth. Arguably, grayscale images don't fit cleanly into that description either. Even OpenGL uses an enum to describe the color format of a texture...

As for the unfortunate issue of having to match over Vec<u8> vs Vec<u16>, I completely agree.

**Comment by dobkeratops** [comment]
(what i've got in mind for further image processing is a multi-dim array with the ability to fold dimensions , e.g.  doing something 'per scanline' is a fold of the outermost axis [row][column][channel], whereas changing pixel packing would be a fold of the last axis; tile interleaving would be like adding more dimensions e.g. turning 256x256 into 16x16 x16x16 - but i don't need that right now.
in my current sourcebase i've got some 2d & 3d array helper classes for landscape manipulation - it might be nice to clean that up and see if it could be applied to image manipulation code

One thing I *might* be doing is automatically generating guesses for 'specular channels' by appliing a filter (clamped difference of gaussians perhaps to pick out ridges), i have an itch to write an auto-tiling utility although preparing textures in GIMP probably better and i could look into script-fu there rather than rust )

**Comment by dobkeratops** [comment]
> Yeah, the issue with just numbers of bits is that there are a bunch of other formats that can't be described that way,

right; and the compressed texture formats are really important to have. so indeed I concede supporting that is essential; there will be a match somewhere. I just wonder if one of the arms could be the 'uncompressed, direct-color' cases . but maybe that itself would be imposing a match-nesting heh.

**Comment by HeroicKatora** [comment]
What seemed to be the result of a discussion on gitter, these two following methods could be useful for `ImageBuffer` to communicate how it uses the underlying container:

```
pub fn dimensions(&self) -> (usize, usize, usize) { 
    (image.width(), image.height(), P::channel_count()) 
} 

pub fn strides(&self) -> (usize, usize, usize) { 
    (image.height()*P::channel_count(), P::channel_count(), 1) 
}
```

Open questions: 
* Should there be another buffer wrapper that takes `T: AsRef<[P::Subpixel]>` and three arbitrary strides, to accomodate cases where the samples are layed out differently in the input container?
* In case that yes, should it maybe even be merged with the current `ImageBuffer`?

**Comment by fintelia** [comment]
This functionality seems completely reasonable to add. `ImageBuffer::dimensions` already exists with a different return type, so it may be better to pick another name. I also don't love using usize in the return types, but u32 or u64 would probably be worse.

**Comment by HeroicKatora** [comment]
The main reason for `usize` is to express that this result is to be used in index calculation, while that is not the main concern of the existing `dimensions`. As a different name, maybe `array_extents` would be better?

**Comment by dobkeratops** [comment]
ok calming down a bit I realise you dont want feature creep (potential future alternate processing library)
but let me update the suggestion to this API - it does mean making a type but named fields explain whats going on

```
     struct InterleavedChannelArray<C:ChannelType> {
         buffer:Vec<C>,
         height:usize,
         width:usize,
         channels:usize,
     }

    impl DynamicImage {  // can't remember specifically.. the type you get from loading
           fn into_interleaved_channel_array<C:ChannelType>()->InterleavedChannelArray<C>;
    }

```

so for my use case i would say ```  image::Load("foo.jpg").into_interleaved_channel_array::<u8>() ````
this allows getting (what I consider to be the most common image representation) in one LOC, it does have a bit of polymorphism i.e. this call could support u8 or f32. it would be a bit more involved to implement i.e some internal polymorphism for conversions for details of u8 or f32 conversion ... alternatively you could just support u8 and leave it to the community to extend if someone wants HDR done this way

the name of the buffer object is open to question .. whats the best balance between explicitness/concision..

you could also embed the strides allowing planar format *however* I would still suggest the underlying call remains 'get_interleaved' getting the specific layout where the channel stride is 1

**Comment by HeroicKatora** [comment]
It's a good point to note that converting a `DynamicImage` into an interleaved array with no associated static color type is potentially less computational work. However, under that point of view the argument for automatically converting sample representation falls a bit flat. It would cause implicit `O(w·h)` processing which would otherwise be avoided. Should we consider immediately consider that  `DynamicImage` needs to represent images with `u16` or other sample types?

The same holds if the conversion would be done as 
`fn into_flat_samples<T>(&self) -> Option<InterleavedChannesl<T>>` 
due to the implicit copy but an alternative is
`fn into_flat_samples<T>(self) -> Result<InterleavedChannels<T>, Self>` 
which would currently only work for `T == u8` anyways, as all underlying representations are of that type. The conversion could also be a trait, so that the container of `InterleavedChannels` can be a `&[P::Subpixel]` as well and both `DynamicImage` and `ImageBuffer` can be converted using the same interfaces.

As a generic output interface that may do arbitrary conversion on samples, the benefit compared to just providing the proposed two additional index-based methods above is much less clear to me.

**Comment by dobkeratops** [comment]
> It would cause implicit O(w·h) processing 

this API is for when you know you want 8bits per channel - if you loaded something else that processing is unavoidable ; in my scenarios i'm loading formats that i know have that;

.. I do accept that eventually i will want to support other formats, but they're not yet supporter here either i.e. DDS compressed textures (and sure the simple image array idea breaks down there, it's a block format

>> , the benefit compared to just providing the proposed two additional index-based methods above is much less clear to me. 

Its just an annoying interface. like many other things in rust. The bloat is coming from having to use the match construct to extract what could just be index information.  The last time I was doing this in C/C++ that's how it worked,  jpg::load_from_file(file, &width,&height,&channels), and it was more pleasant than this


Perhaps I  just made a mistake here thinking I can use someone elses library to load an image.


please, prove my misconceptions wrong.

**Comment by HeroicKatora** [comment]
Some additional thought: `InterleavedChannels`, or any stride-based matrix representation, has an implicit invariant. Namely that the strides will only generate valid indices if multiplied and index with valid indices. Checking this invariant would require the member to be non-public so we need to be explicit about this not being the case, and provide a check method.

**Comment by HeroicKatora** [comment]
> oh jesus christ...

Please read the full argument and stay on-topic. Again, Rust is more oriented towards stability.

> C/C++ that's how it worked, jpg::load_from_file(file, &width,&height,&channels)

If you want to play example, compared the `pnm` loader from [`image::pnm::decoder`](https://github.com/PistonDevelopers/image/blob/master/src/pnm/decoder.rs) and [`imagemagick::pnm`](https://github.com/ImageMagick/ImageMagick/blob/83d74de0d48e0b5116612124233bef1d4713eb82/coders/pnm.c) and you tell me which is actually maintainable or analyzable by anyone but the original author.

So I must ask you to stay professional. This is the third time and many of your argument have already been considered and there was a consensus that some requests are reasonable. If you can not properly handle an technical evaluation of pros and cons, I see no way that this continues in a productive manner (Meaning I will close this issue and we continue adding the agreed upon improvements fully internally.).

**Comment by dobkeratops** [comment]
> that there should be two different interfaces, one which does the costly conversion, and one which has a guarantee of not doing it. 

thats fine; my suggestions are orthogonal to the other side of things where I understand you may wrap many different types of image with underlying representations, and you want to query exactly what was loaded.

I think whats gone wrong here for me is..

... i had some nice simple, working code in C++ 5 years ago, which loaded TGAs and JPGs, into an array. then I came to Rust, used FFI to grab them. Then I thought "hey, lets get into the spirit of the rust community and use an existing crate and rust code" ... and here I am jumping through hoops with layers of wrappers that bloat and confuse my code to do the same thing, and I pretty much have to make another wrapper function to make the library pleasant to use.

sorry to get frustrated and lash out.. but it must be possible to streamline

**Comment by HeroicKatora** [comment]
> then I came to Rust, used FFI to grab them.

I see how that would easily work in your single own project. But designing an interface that scales to a more general setting is not as easy as exposing an ffi function pointer, especially when you have to consider the Rust stance on undefined behaviour and memory safety.

**Comment by dobkeratops** [comment]
putting the data in a Vec<> achieves memory safety. 

thats no excuse here.

**Comment by dobkeratops** [comment]
ok. so if i just want to load an array of pixels i have to write  my own library rather than re-use community libraries, got it.

although in C++ i could use an existing library, 

hmmm.

**Comment by HeroicKatora** [comment]
If you feel like that is your sole choice, there is nothing stopping you. That's why this project is open source. @bvssvni  Can I close this?

**Comment by dobkeratops** [comment]
> If you want to play example, compared the pnm loader from image::pnm::decoder and imagemagick::pnm and you tell me which is actually maintainable or analyzable by anyone but the original author.

nice straw man, comparing Rust to *C* instead of *C++*

**Comment by bvssvni** [comment]
@HeroicKatora there was this heavy debate some years ago when @dobkeratops , me and others wanted more gamedev ergonomic features in Rust, since we felt that the common language designer weren't thinking about those use cases that costed programmers hours and hours in long term projects. Gamedevs were really hoping Rust would move more in their direction, but they saw focus on safety as an overarching conflicting goal with their needs.

The lasting impression from that discussion was that some people did not really listened, or even made fun of these concerns (however, taken seriously by some very good programmers), but in the discussion a lot of interesting ideas came up. Jonathan Blow started working on Jai, I got interested in language design and made Dyon to experiment with some ideas, and @dobkeratops worked on his own language, written in C++.

I think that's the underlying issue here, which got very little to do with how you work with pixels.

@dobkeratops I have been thinking of some ideas of language design lately, starting a new thread here: https://github.com/PistonDevelopers/piston/issues/1253

**Comment by HeroicKatora** [comment]
I had a growing suspicion that this problem was rooted much deeper, although I did not know of the lengthy prior legacy. However, there are alternatives to approaching these concerns, alternatives which are more healthy to the state of this library than approach its design from the perspective of a different language. Apparantly `ffi` has already worked successfully in parts.

And there are more pressing issues that actually limit functionality instead of revolving around mostly sparing 5 lines of wrapper code. I tried my best to extract those suggestions from the stylistic questions, hoping that this would at least partially solve those as well by embedding some documentation assumptions in runtime values.

**Comment by bvssvni** [comment]
@HeroicKatora I agree that we should focus on stability.

**Comment by dobkeratops** [comment]
ok i'll apologise for outbursts recently, I was lashing out. I have some broad frustrations and you might have just got me on the wrong day. I have found Image:: useful.

I might think about making an 'ImageArray' crate (somewhere between simple image handling, voxel grids, and interface to machinelearning stuff i do intend to dip back into later),  but i'll probably run into the issue that its hard to get anyone to notice. it would be nice to give back to the community. but any time i've tried I find ideas clash very quickly . it's nearly impossible to do.  so i'll probably just roll an ad-hoc thing. It's still nice to compare ideas.
